design/src/rfcs/rfc0018_logging_sensitive.md

```diff
@@ -0,0 +1,407 @@
+
+Each of these configurable parts must therefore be logged cautiously.
+
+### Scope and Guidelines
+
+It would be unfeasible to forbid the logging of sensitive data all together using the type system. With the current API, the customer will always have an opportunity to log a request containing sensitive data before it enters the `Service<Request<B>>` that we provide to them.
+
+```rust
+// The API provides us with a `Service<Request<B>>`
+let app: Router = OperationRegistryBuilder::default().build().expect("unable to build operation registry").into();
+
+// We can `ServiceExt::map_request` log a request with potentially sensitive data
+let app = app.map_request(|request| {
+        info!(?request);
+        request
+    });
+```
+
+A more subtle violation of the specification may occur when the customer enables verbose logging - a third-party dependency might simply log data marked as sensitive, for example `tokio` or `hyper`.
+
+These two cases illustrate that `smithy-rs` can only prevent violation of the specification in a restricted scope - logs emitted from generated code and the runtime crates. A `smithy-rs` specific guideline should be available to the customer which outlines how to avoid violating the specification in areas outside of our control.
+
+### Routing
+
+The sensitivity and HTTP bindings are declared within specific structures/operations. For this reason, in the general case, it's unknowable whether or not any given part of a request is sensitive until we determine which operation is tasked with handling the request and hence which fields are bound. Implementation wise, this means that any middleware applied _before_ routing has taken place cannot log anything sensitive without performing routing logic itself.
+
+Note that:
+
+- We are not required to deserialize the entire request before we can make judgments on what data is sensitive or not - only which operation it has been routed to.
+- We are permitted to emit logs prior to routing when:
+  - they contain no potentially sensitive data, or
+  - the request failed to route, in which case it's not subject to the constraints of an operation.
+
+### Runtime Crates
+
+The crates existing in `rust-runtime` are not code generated - their source code is agnostic to the specific model in use. For this reason, if such a crate wanted to log potentially sensitive data then there must be a way to conditionally toggle that log without manipulation of the source code. Any proposed solution must acknowledge this concern.
+
+## Proposal
+
+This proposal serves to honor the sensitivity specification via code generation of a logging middleware which is aware of the sensitivity, together with a developer contract disallowing logging potentially sensitive data in the runtime crates. An internal and external guideline should be provided in addition to the middleware.
```

Comment:
> internal and external guideline

Why would there be a need for two guidelines?

Reply:
I read the sections for the guidelines below and still don't understand why we need two guidelines. Aren't all the bullet points in those sections useful for both internal and external developers?

Reply:
Good point.

To clarify, by external developers I mean someone who has a model and wants to use `smithy-rs` to construct a server. In this case the external developer would be composing/initializing `Subscribers`, whereas the internal developer wouldn't be. This means internal developers, while just using `tracing` crate, aren't concerned with:

> - Sensitive data leaking from third-party dependencies.
> - Sensitive data leaking from middleware applied to the `Router`.

There might also be a slight difference in emphasis about certain things:

- The internal developer needs to very familiar with the concept of "potentially sensitive data" because, outside of codegen, they cannot make assumptions about the model.
- The external developer knows the model they're using and therefore knows exactly which data is sensitive.

But yeah, these differences are probably not large enough to warrant a separate guide.